src/L1/rollup/ScrollChain.sol

```diff
@@ -438,149 +369,67 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
-            _publicInputHash = keccak256(
-                abi.encodePacked(
-                    layer2ChainId,
-                    _prevStateRoot,
-                    _postStateRoot,
-                    _withdrawRoot,
-                    _dataHash,
-                    _blobDataProof[0:64],
-                    _blobVersionedHash
-                )
-            );
-        }
-
-        // load version from batch header, it is always the first byte.
-        uint256 batchVersion = BatchHeaderV0Codec.getVersion(batchPtr);
-        // verify batch
-        IRollupVerifier(verifier).verifyAggregateProof(batchVersion, _batchIndex, _aggrProof, _publicInputHash);
-
-        // Pop finalized and non-skipped message from L1MessageQueue.
-        uint256 _totalL1MessagesPoppedOverall = BatchHeaderV0Codec.getTotalL1MessagePopped(batchPtr);
-        _popL1MessagesMemory(
-            BatchHeaderV1Codec.getSkippedBitmapPtr(batchPtr),
-            _totalL1MessagesPoppedOverall,
-            BatchHeaderV0Codec.getL1MessagePopped(batchPtr)
-        );
-
-        _afterFinalizeBatch(_totalL1MessagesPoppedOverall, _batchIndex, _batchHash, _postStateRoot, _withdrawRoot);
-    }
-
     /// @inheritdoc IScrollChain
+    /// @dev All batches in the given bundle should have the same version and version <= 4 or version >= 6.
     function finalizeBundleWithProof(
-        bytes calldata _batchHeader,
-        bytes32 _postStateRoot,
-        bytes32 _withdrawRoot,
-        bytes calldata _aggrProof
+        bytes calldata batchHeader,
+        bytes32 postStateRoot,
+        bytes32 withdrawRoot,
+        bytes calldata aggrProof
     ) external override OnlyProver whenNotPaused {
-        if (_postStateRoot == bytes32(0)) revert ErrorStateRootIsZero();
-
-        // compute pending batch hash and verify
+        // actions before verification
         (
-            uint256 batchPtr,
-            bytes32 _batchHash,
-            uint256 _batchIndex,
-            uint256 _totalL1MessagesPoppedOverall
-        ) = _loadBatchHeader(_batchHeader);
+            uint256 version,
+            bytes32 batchHash,
+            uint256 batchIndex,
+            uint256 totalL1MessagesPoppedOverall,
+            uint256 prevBatchIndex
+        ) = _beforeFinalizeBatch(batchHeader, postStateRoot);
 
-        // retrieve finalized state root and batch hash from storage to construct the public input
-        uint256 _finalizedBatchIndex = lastFinalizedBatchIndex;
-        if (_batchIndex <= _finalizedBatchIndex) revert ErrorBatchIsAlreadyVerified();
+        uint256 euclidForkBatchIndex = initialEuclidBatchIndex;
+        // Make sure we don't finalize v4, v5 and v6 batches in the same bundle, that
+        // means `batchIndex < euclidForkBatchIndex` or `prevBatchIndex >= euclidForkBatchIndex`.
+        if (prevBatchIndex < euclidForkBatchIndex && euclidForkBatchIndex <= batchIndex) {
+            revert ErrorFinalizePreAndPostEuclidBatchInOneBundle();
```

Comment:
Well if we try to do this the proof verification should never pass. But we can keep this as a sanity check if you prefer.